runtime/gc_modron_startup/mgcalloc.cpp

```diff
@@ -545,6 +545,27 @@ J9AllocateIndexableObject(J9VMThread *vmThread, J9Class *clazz, uint32_t numberO
 		/* The hook could release access and so the object address could change (the value is preserved).  Since this
 		 * means the hook could write back a different value to the variable, it must be a valid lvalue (ie: not cast).
 		 */
+
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+		GC_ArrayletObjectModel::ArrayLayout layout = extensions->indexableObjectModel.getArrayLayout((J9IndexableObject*)objectPtr);
+		if (extensions->indexableObjectModel.isDoubleMappingEnabled() && extensions->isVLHGC() && (layout != GC_ArrayletObjectModel::InlineContiguous)) {
+			Assert_MM_true(layout == GC_ArrayletObjectModel::Discontiguous);
+			UDATA arrayletLeafCount = extensions->indexableObjectModel.numArraylets((J9IndexableObject*)objectPtr);
+			UDATA sizeInElements = extensions->indexableObjectModel.getSizeInElements((J9IndexableObject*)objectPtr);
+			Assert_MM_true(sizeInElements == 0 || arrayletLeafCount > 0);
+
+			if ((arrayletLeafCount > 1) && (sizeInElements > 0)) {
+				extensions->doubleMapArraylets(env, objectPtr);
```

Comment:
1) I'm not sure about this, but maybe all the condition checks should be inside the doubleMapArraylets call. I don't want to make J9AllocateIndexableObject too fat (well, it already is).
2) I wonder if there is another spot deeper in the alloc path where we already know that these conditions are met, and call doubleMapArraylets  from there.
3) Again, not sure if doubleMapArraylets  should be in GC Extensions. Typically we have little functional code there, related to init/freee/startup/shutdown. The method could be in this file (if we don't find another sport to call it from).

@dmitripivkine, thoughts?

Reply:
This seems like a good spot to call it from, right after arrayoid is populated: MM_IndexableObjectAllocationModel::layoutDiscontiguousArraylet